crates/pinned_pool/examples/pp_trait_objects.rs

```diff
@@ -0,0 +1,211 @@
+//! Trait object usage with `PinnedPool`.
+//!
+//! This example demonstrates how to use trait objects with `PinnedPool`, showing:
+//! * Storing different types that implement the same trait
+//! * Converting pinned references to trait object references
+//! * Using trait methods on pooled items
+
+use std::fmt::Display;
+
+use pinned_pool::PinnedPool;
+
+// Define a trait that we'll use for trait objects.
+trait Drawable {
+    fn draw(&self) -> String;
+    fn area(&self) -> f64;
+}
+
+// Different types implementing the trait.
+#[derive(Debug)]
+struct Circle {
+    radius: f64,
+}
+
+impl Drawable for Circle {
+    fn draw(&self) -> String {
+        format!("Drawing circle with radius {}", self.radius)
+    }
+
+    fn area(&self) -> f64 {
+        std::f64::consts::PI * self.radius * self.radius
+    }
+}
+
+impl Display for Circle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Circle(radius={})", self.radius)
+    }
+}
+
+#[derive(Debug)]
+struct Rectangle {
+    width: f64,
+    height: f64,
+}
+
+impl Drawable for Rectangle {
+    fn draw(&self) -> String {
+        format!("Drawing rectangle {}x{}", self.width, self.height)
+    }
+
+    fn area(&self) -> f64 {
+        self.width * self.height
+    }
+}
+
+impl Display for Rectangle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Rectangle({}x{})", self.width, self.height)
+    }
+}
+
+fn main() {
+    println!("PinnedPool Trait Object Example");
+    println!("===============================");
+    println!();
+
+    // Method 1: Store trait objects directly in the pool.
+    println!("Method 1: Storing trait objects directly");
+    println!("----------------------------------------");
+
+    let mut shape_pool = PinnedPool::<Box<dyn Drawable>>::new();
```

Comment:
While technically valid, this just reduces the trait object part of the problem to storage of `Box` - this is not really anything specific to PinnedPool, you can always use a boxed trait object with collections. We should instead store a specific type and only treat it as a trait object after taking it from storage, to show that this works. Refactor this example to just use a `PinnedPool<Rectangle>`, for example, from which we take `&dyn Drawable` references.